src/TestCase.ts

```diff
@@ -0,0 +1,241 @@
+
+      case "list":
+        return target.elements.map(this.extractTargetKeys, this).flat();
+
+      case "range":
+        return this.extractPrimitiveTargetKeys(target.start, target.end);
+
+      default:
+        return [];
+    }
+  }
+
+  extractTargetedDecorations(targets: Target[], navigationMap: NavigationMap) {
+    if (!navigationMap) {
+      return {};
+    }
+
+    const decorationRanges = navigationMap.serializeRanges();
+    const targetedDecorations: DecorationRanges = {};
+    const targetKeys = targets.map(this.extractTargetKeys, this).flat();
+    targetKeys.forEach((key) => {
+      targetedDecorations[key] = decorationRanges[key];
+    });
+    return targetedDecorations;
+  }
+
+  isThatMarkTargeted() {
+    return this.targets.some(
+      (target) => target.type === "primitive" && target.mark.type === "that"
+    );
+  }
+
+  static async getSnapshot(
+    thatMark: SelectionWithEditor[]
+  ): Promise<TestCaseSnapshot> {
+    const activeEditor = vscode.window.activeTextEditor!;
+    return {
+      document: activeEditor.document.getText(),
+      selections: activeEditor.selections.map(serializeSelection),
+      visibleRanges: activeEditor.visibleRanges.map(serializeRange),
+      clipboard: await Clipboard.readText(),
+      thatMark: thatMark.map((mark) => serializeSelection(mark.selection)),
+    };
+  }
+
+  async getSnapshot(): Promise<TestCaseSnapshot> {
+    return await TestCase.getSnapshot(this.context.thatMark.get());
+  }
+
+  async saveSnapshot() {
+    const snapshot = await this.getSnapshot();
+
+    if (!["copy", "paste"].includes(this.command.actionName)) {
+      snapshot.clipboard = "";
+    }
+
+    if (!["fold", "unfold"].includes(this.command.actionName)) {
+      snapshot.visibleRanges = [];
+    }
+
+    if (this.initialState == null && !this.isThatMarkTargeted()) {
+      snapshot.thatMark = [];
+    }
+
+    if (this.initialState == null) {
+      this.initialState = snapshot;
+    } else if (this.finalState == null) {
+      this.finalState = snapshot;
+    } else {
+      throw Error("Both snapshots already taken");
+    }
+
+    return snapshot;
+  }
+
+  toYaml() {
+    if (this.initialState == null || this.finalState == null) {
+      throw Error("Two snapshots must be taken before serializing");
+    }
+    const fixture: TestCaseFixture = {
+      talonCommand: this.talonCommand,
+      command: this.command,
+      languageId: this.languageId,
+      targets: this.targets,
+      marks: this.marks,
+      initialState: this.initialState,
+      finalState: this.finalState,
+      returnValue: this.returnValue,
+    };
+    return yaml.dump(fixture, { noRefs: true, quotingType: '"' });
+  }
+
+  async writeFixture(filename: string) {
+    if (filename === "") {
+      throw new Error("Filename required");
+    }
+
+    const fixture = this.toYaml();
+    const workspacePath = vscode.workspace.workspaceFolders?.[0].uri.path;
+    let document;
+
+    if (workspacePath) {
```

Comment:
when will this one be null?

Reply:
If we are in a fresh, untitled window, ex. `File > New Window`, or if only a file is opened, not a folder.